db.go

```diff
@@ -3,30 +3,191 @@ package fury
 // DB object consists of DB connection pool and configuration
 // 	Use Connect(config *Configuration) to create new instance of this struct.
 type DB struct {
-	connectionPool
+	ConnectionPooler
 	config *Configuration
-	*Query
+	query  *Query
 }
 
 // Connect to database, instantiate DB struct for querying to database.
 func Connect(config *Configuration) (*DB, error) {
 	config.initialize()
 
-	newConnPool, err := newConnectionPool(config)
-
+	newConnPool, err := NewConnectionPool(config)
 	if err != nil {
 		return nil, err
 	}
 
 	db := &DB{
-		connectionPool: newConnPool,
-		config:         config,
+		ConnectionPooler: newConnPool,
+		config:           config,
 	}
 
 	return db, nil
 }
 
-// Model function to set table from struct
-func (db *DB) Model(model interface{}) {
-	return
+func (db *DB) clone(model interface{}) (*DB, error) {
+	q, err := NewQuery(model)
+	if err != nil {
+		return nil, err
+	}
+
+	return &DB{
+		ConnectionPooler: db.ConnectionPooler,
+		config:           db.config,
+		query:            q,
+	}, nil
+}
+
+// First method return first record ordered by primary key
+func (db *DB) First(model interface{}, opts ...QueryOption) error {
+	opts = append(opts, Limit(1))
+	return db.Find(model, opts...)
+}
+
+// Find method return all record queried with specified conditions
+func (db *DB) Find(model interface{}, opts ...QueryOption) error {
+	newDB, err := db.clone(model)
+	if err != nil {
+		return err
+	}
+
+	for _, opt := range opts {
+		_, err := opt(newDB.query)
+		if err != nil {
+			return err
+		}
+	}
+
+	return newDB.executeSelectQuery()
+}
+
+// Insert query method
+func (db *DB) Insert(model interface{}, opts ...QueryOption) error {
+	newDB, err := db.clone(model)
+	if err != nil {
+		return err
+	}
+
+	for _, opt := range opts {
+		_, err := opt(newDB.query)
+		if err != nil {
+			return err
+		}
+	}
+
+	return newDB.executeInsertQuery()
+}
+
+// Update query method
+func (db *DB) Update(model interface{}, opts ...QueryOption) error {
+	newDB, err := db.clone(model)
+	if err != nil {
+		return err
+	}
+
+	for _, opt := range opts {
+		_, err := opt(newDB.query)
+		if err != nil {
+			return err
+		}
+	}
+
+	return newDB.executeUpdateQuery()
+}
+
+// Delete query method
+func (db *DB) Delete(model interface{}, opts ...QueryOption) error {
+	newDB, err := db.clone(model)
```

Comment:
setiap kali operation `db.clone` setelah selesai melakukan database, kenapa connection databasenya tidak di close?

Reply:
`db.clone` return copy dari struct `DB` dengan `ConnectionPooler` dan `Configuration` yang sama, tapi setiap clone dia bakalan create `Query` baru untuk generate query baru. Nah karena `ConnectionPooler` sama jadi cuma perlu di-close saat terakhir program-nya jalan. Waktu create connection biasa atau request biasa ngga perlu close `ConnectionPooler`-nya

Reply:
i see, karena menggunakan pointer of `ConnectionPoller` yah. ini gmn jika ada dua user melakukan operation secara bersamaan akan terjadi `data race`. coba baca ini 

https://notes.shichao.io/gopl/ch9/

Reply:
Kalau di bagian data race di database-nya masih, karena ini memang masih implementasi basic-nya belum menambahkan transaction. Tapi harusnya untuk di code-nya sudah kan ya? karena ini clone dulu dan `DB` yang digunakan dari `database/sql` itu concurrent-safe. Tapi kalau di code untuk query-nya kalau misalkan dibuat parallel kemungkinan bisa data race karena ada banyak slice dan map yang belum diberi lock untuk aksesnya

Reply:
ini kita pisahkan test pada database dll. untuk library nya sendiri hal ini masih dapat terjadi data race.  jika function clone ini tidak ambil pointer of struct tapi copy value nya saja bagaiamana menurutmu?  data-race ini memang sulit untuk ditemukan. jika menurut kamu bagian ini cukup sulit diskip saja dan fix bagian lain gpp.

nanti lanjut ke task onboarding selanjutnya.

Reply:
Sebenernya cuma masih bingung data race-nya ada di bagian mana. Kalau misalkan copy value dari `ConnectionPooler` itu nanti terhitungnya dua connection pool ke database? Dan kalau begitu, apakah proses untuk copy akan open connection lagi ke database?

Untuk bagian ini sepertinya mau coba untuk di-fix dulu, gpp biar sekalian belajar. Untuk task selanjutnya boleh sekalian kerjain aja? Yang penting nanti waktu code bagian repo-nya terakhir.

Reply:
concern saya karena db clone itu share pointer of connection pooler. jadi jika ada dua user yang melakukan operation secara bersamaan itu akan terjadi data race. 

contohnya mungkin seperti ini. di db semuanya berjalan dengan baik tapi pada library data yang dikembalikan tidak benar, data race mungkin terjadi di 1 dari 1000 operation tapi ini akan untuk me-tracenya ketika terjadi.

![Screenshot from 2019-06-30 00-18-02](https://user-images.githubusercontent.com/7883991/60387399-9ed0ca00-9acc-11e9-91b1-e3b17483d11a.png)

untuk memulai mengerti apa itu data-race dan bagaimana mengatasinya bisa mulai membaca ini:
- https://golang.org/doc/articles/race_detector.html
- https://medium.com/@val_deleplace/does-the-race-detector-catch-all-data-races-1afed51d57fb

kita berlajar bareng saja, kadang saya pun sulit untuk mengatasi data-race ini :)

Reply:
Kalau sesuai contoh di atas, sepertinya itu masalahnya ada di bagian lock database-nya. Jika level akses read yang dipilih bisa read uncommitted data, maka operasi read akan melakukan dirty read. Tapi kalau yang dipilih level yang lebih menjaga data maka harusnya tidak terjadi seperti contoh tersebut.

Saya juga sekalian lagi cari sih untuk case bagaimana kemungkinan data race-nya akan terjadi. Walaupun di [dokumentasi library-nya](https://golang.org/pkg/database/sql/#DB ) tertulis:

> DB is a database handle representing a pool of zero or more underlying connections. It's safe for concurrent use by multiple goroutines.
dan di link ini https://notes.shichao.io/gopl/ch9/#chapter-9-concurrency-with-shared-variables :

> Indeed, concurrency-safe types are the exception rather than the rule, so you should access a variable concurrently only if the documentation for its type says that this is safe.

Masih ada kurang aman untuk concurrency ya? Kalau seperti itu berarti perlu ada mutex lagi atau copy value. Tapi dengan seperti itu pasti menambah overhead. Untuk copy value, saya harus coba dulu sih, ini nantinya jadi connection baru atau tidak, kalau terhitung connection baru kemungkinan overhead bertambah banyak karena akan open dan close connection terus.

Reply:
iya jika kamu pakai struct dari library `database/sql`. tapi kamu menghandle DB connection pooler sendiri pada library mu jadi kamu perlu meyakinkan bahwa ini safe for concurrent.

jika kamu lihat source code dari `database/sql` dia menggunakan mutex.Sync untuk mengatasi concurrent, lihat disini https://golang.org/src/database/sql/sql.go?s=9571:11341#L340.

iya kamu bisa tambahkan mutext dan lakukan test apakah overhead nya worhty atau tidak.

Reply:
Bukannya kalau implementasi sekarang sama saja dengan menggunakan struct dari `database/sql`. Karena sebenarnya `ConnectionPooler` hanya contract untuk connection ke DB yang mana implementasinya, termasuk untuk concurrency, di-handle oleh apapun yang mengimplementasikan `ConnectionPooler`. Dalam kasus ini yang mengimplementasikan adalah struct dari `database/sql`, jadi bukannya sebenernya sama saja ya? Dan kalau ditambahkan mutex lagi jadi dua kali. Kalau ditambahkan juga jadinya malah lock seluruh resource, bukan inti resource untuk connection ke DB yang sudah di-handle oleh library `database/sql`. Ada yang salah ngga dari pemahaman saya seperti itu?

Reply:
owh kamu hanya bungkus database/sql nya yah. i see, saya pikir kamu handle connection pooler sendiri, line code ini https://github.com/nandaryanizar/fury/blob/feature/query/connection_pool.go#L32 sepertinya terlewat oleh saya. sip